### PR TITLE
Feature/3 - Onboarding 

### PR DESCRIPTION
**Pull request by ghost**
- New Page to the Onboarding initial step
- Routes file to map the pages
- HomePage created to simulate the Onboarding process

JIRA Items:

- Task 3
- Subtasks 17 and 18

### CODE REVIEW

File: src/pages/home/HomePage.tsx [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import React from 'react'
+import logo from '../../logo.svg';
+
+export default class HomePage extends React.Component {
+    render() {
+        return (
+            <header className="App-header">
```
Comment (arctouch-marceloaquino) [inline src/pages/home/HomePage.tsx]: We don't need this page anymore, our first page (on first access) will be the Onboarding page and after that will be the Forecast page.

File: src/pages/onboarding/OnboardingPage.css [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    background-size: cover;
+}
+
+.OnboardingPage-skipButton {
+    width: 44px;
+    height: 27px;
+    font-family: Avenir;
+    font-size: 20px;
+    font-weight: 300;
+    font-style: normal;
```
Comment (arctouch-marceloaquino) [inline src/pages/onboarding/OnboardingPage.css]: There are some classes here that could be merged into a single one.

File: src/pages/onboarding/OnboardingPage.tsx [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    onboardingService: OnboardingService
+
+    constructor(props: Props) {
+        super(props)
+        this.moveToNextStep = this.moveToNextStep.bind(this)
+        this.finishOnboarding = this.finishOnboarding.bind(this)
+        this.onboardingService = new OnboardingService()
+    }
+
+    state = {
```
Comment (arctouch-marceloaquino) [inline src/pages/onboarding/OnboardingPage.tsx]: I'd suggest to move the state initialize to inside the constructor.

File: src/pages/onboarding/OnboardingPage.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+    render() {
+        return (<div className="OnboardingPage">
+            <span className="OnboardingPage-skipButton" onClick={this.finishOnboarding}>Skip</span>
+            <Carousel
+                className="OnboardingPage-carousel"
+                renderCenterRightControls={null}
+                renderCenterLeftControls={null}
+                slideIndex={this.state.currentIndexCarousel}
+                afterSlide={currentIndexCarousel => this.setState({ currentIndexCarousel })}
```
Comment (arctouch-marceloaquino) [inline src/pages/onboarding/OnboardingPage.tsx]: I'm not sure if it's a good practice to set state value this way. Probably will be better to call a function that will do that.

**Review by arctouch-marceloaquino — CHANGES_REQUESTED** [review]
There are some issues to be fixed.

File: src/pages/onboarding/OnboardingPage.tsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    }
+]
+
+export default class OnboardingPage extends React.Component<Props, State> {
+
+    onboardingService: OnboardingService
+
+    constructor(props: Props) {
+        super(props)
+        this.moveToNextStep = this.moveToNextStep.bind(this)
```
Comment (arctouch-jonathansouza) [inline src/pages/onboarding/OnboardingPage.tsx]: We are not using binding to invoke callbacks to React. We'd prefer using arrow functions (`() => { ... }`), that makes the code cleaner and void these boilerplate code.

File: src/pages/onboarding/OnboardingPage.tsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        return (
+            <div className="OnboardingPage-content">
+                <span className="OnboardingPage-stepTitle">{OnboardingSteps[this.state.currentIndexCarousel].title}</span>
+                <span className="OnboardingPage-stepDescription">{OnboardingSteps[this.state.currentIndexCarousel].description}</span>
+            </div>
+        )
+    }
+
+    renderNextButton(): React.ReactNode {
+
```
Comment (arctouch-jonathansouza) [inline src/pages/onboarding/OnboardingPage.tsx]: This blank line can be deleted.

File: src/pages/onboarding/OnboardingPage.tsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        this.finishOnboarding = this.finishOnboarding.bind(this)
+        this.onboardingService = new OnboardingService()
+    }
+
+    state = {
+        currentIndexCarousel: 0
+    }
+
+    moveToNextStep() {
+        if (this.state.currentIndexCarousel < (OnboardingSteps.length - 1)) {
```
Comment (arctouch-jonathansouza) [inline src/pages/onboarding/OnboardingPage.tsx]: For the cases like this where we have a complex boolean expression inside an `if`, I'd suggest us to extract the expression inside a function, that would make the code easier for the developer that is reading it.

File: src/pages/onboarding/OnboardingPage.tsx [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+                <span className="OnboardingPage-stepDescription">{OnboardingSteps[this.state.currentIndexCarousel].description}</span>
+            </div>
+        )
+    }
+
+    renderNextButton(): React.ReactNode {
+
+        let nextButtonText = "Next"
+
+        if (this.state.currentIndexCarousel === (OnboardingSteps.length - 1))
```
Comment (arctouch-jonathansouza) [inline src/pages/onboarding/OnboardingPage.tsx]: Same above

File: src/pages/onboarding/OnboardingPage.tsx [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            this.finishOnboarding()
+        }
+    }
+
+    finishOnboarding() {
+        this.onboardingService.setOnboardingFinished()
+        //TODO: Go to the Fresh Install Page
+    }
+
+    renderOnboardingSteps(): JSX.Element[] {
```
Comment (arctouch-jonathansouza) [inline src/pages/onboarding/OnboardingPage.tsx]: Usually, Typescript is able to infer the function return type, so we can delete these return definitions. We can use them only when we need to have something different from what is inferred from TS.

File: src/pages/home/HomePage.tsx (hunk 1 above)
Comment (arctouch-alyssonlopes) [reply src/pages/home/HomePage.tsx]: HomePage was created to simulate the Onboarding logic. It's required for the tests.

File: src/pages/onboarding/OnboardingPage.tsx [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    title: string,
+    description: string,
+    imageSrc: string
+}
+
+const OnboardingSteps: OnboardingStep[] = [
+    {
+        title: 'Welcome to ArcWeather',
+        description: 'Your new favorite weather app, with all the features you need to have the best information available.',
+        imageSrc: 'assets/onboarding-image-01.png'
```
Comment (arctouch-jonathansouza) [inline src/pages/onboarding/OnboardingPage.tsx]: Are these onboarding files on `/assets` or `/images`?

**Review by arctouch-jonathansouza — CHANGES_REQUESTED** [review]

File: src/pages/onboarding/OnboardingPage.tsx (hunk 10 above)
Comment (arctouch-alyssonlopes) [reply src/pages/onboarding/OnboardingPage.tsx]: These onboarding images are on /assets public static folder. 
What's the right way to do in this case, @arctouch-jonathansouza?

File: src/pages/onboarding/OnboardingPage.tsx [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+      <img
+        key={index}
+        src={step.imageSrc}
+        alt={step.title}
+        title={step.title}
+      />
+    ));
+  }
+
+  renderStepText(): React.ReactNode {
```
Comment (arctouch-jonathansouza) [inline src/pages/onboarding/OnboardingPage.tsx]: This return type can be omitted

File: src/pages/onboarding/OnboardingPage.tsx [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+          {OnboardingSteps[this.state.currentIndexCarousel].title}
+        </span>
+        <span className="OnboardingPage-textContent OnboardingPage-stepDescription">
+          {OnboardingSteps[this.state.currentIndexCarousel].description}
+        </span>
+      </div>
+    );
+  }
+
+  renderNextButton(): React.ReactNode {
```
Comment (arctouch-jonathansouza) [inline src/pages/onboarding/OnboardingPage.tsx]: Same here

File: src/pages/onboarding/OnboardingPage.tsx [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+        <img
+          className="OnboardingPage-arrowToRight"
+          src={ArrowToRightIcon}
+          alt="Arrow to Right"
+        />
+      </div>
+    );
+  }
+
+  renderRedirect(): React.ReactNode {
```
Comment (arctouch-jonathansouza) [inline src/pages/onboarding/OnboardingPage.tsx]: Same

**Review by arctouch-jonathansouza — DISMISSED** [review]
Left some minor comments. Way to go! 💪

**Review by arctouch-caiodoneda — DISMISSED** [review]

File: src/Routes.tsx [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+        <Switch>
+            <Route exact path='/' component={getMainPage()} />
+            <Route exact path='/fresh-install' component={FreshInstallPage} />
+        </Switch>
+    </HashRouter>
+)
+
+const getMainPage = () => {
+    const onboardingService = new OnboardingService()
+    return onboardingService.isOnboardingFinished() ? FreshInstallPage : OnboardingPage
```
Comment (arctouch-marceloaquino) [inline src/Routes.tsx]: @arctouch-jonathansouza @arctouch-caiodoneda Here we have a situation that me and @arctouch-alyssonlopes were discussing to see if we reach somewhere. Basically the question is "Does make sense to redirect the user directly to Fresh Install page if he already passed through onboarding page but closed the app ?"

**Review by arctouch-marceloaquino — DISMISSED** [review]
Minor issues but LGTM. Good job.

File: src/Routes.tsx (hunk 14 above)
Comment (arctouch-jonathansouza) [reply src/Routes.tsx]: Yes, the onboarding page should be displayed only once

**Review by arctouch-jonathansouza — APPROVED** [review]

**Review by arctouch-marceloaquino — APPROVED** [review]

**Pull request merged**